src/track/track.cpp

```diff
@@ -1103,6 +1103,44 @@ void Track::removeCuesOfType(mixxx::CueType type) {
     }
 }
 
+void Track::swapHotcues(int from, int to) {
+    auto locked = lockMutex(&m_qMutex);
+    qWarning() << "     Track: swapHotcues" << from << "to" << to;
+    CuePointer pCueA;
+    CuePointer pCueB;
+    for (const CuePointer& pCue : m_cuePoints) {
+        if (pCueA && pCueB) {
+            qWarning() << "     > found both, break";
+            break;
+        }
+        if (pCue->getType() != mixxx::CueType::HotCue) {
+            continue;
+        }
+        if (pCue->getHotCue() == from) {
+            qWarning() << "     > found FROM" << from;
+            pCueA = pCue;
+            continue;
+        }
+        if (pCue->getHotCue() == to) {
+            qWarning() << "     > found TO" << to;
+            pCueB = pCue;
+            continue;
+        }
+    }
+    if (!pCueA && !pCueB) {
+        return;
+    }
+    if (pCueA) {
+        qWarning() << "     >> move" << from << "to" << to;
+        pCueA->setHotCue(to);
+    }
+    if (pCueB) {
+        qWarning() << "     >> move" << to << "to" << from;
+        pCueB->setHotCue(from);
+    }
+    emit cuesUpdated();
+}
```

Comment:
I removed the cuetype restriction as I didn't understand why it would be necessary. Wdyt?
```suggestion
void Track::swapHotcues(int a, int b) {
    auto locked = lockMutex(&m_qMutex);
    // REVIEW COMMENT: finding a specific cue by its index could also be a utility member function
    auto is_cue_n = [](int n) {
        return [n](const CuePointer& pCue){
            return pCue && pCue->getHotCue() == n;
        }
    };
    auto cueAIt = std::find_if(m_cuePoints.begin(), m_cuePoints.end(), is_cue_n(a));
    if (cueAIt == m_cuePoints.end()) {
        qWarning() << "could not find Cuepoint" << a << "for swapping with" << b;
        return;
    }
    auto cueBIt = std::find_if(m_cuePoints.begin(), m_cuePoints.end(), is_cue_n(b));
    if (cueBIt == m_cuePoints.end()) {
        qWarning() << "could not find Cuepoint" << b << "for swapping with" << a;
    }
    // we already know the pointers are non-null from is_cue_n
    // now we deliberately set cueB to a and cueA to b to perform the
    // actual swap. 
    (*cueBIt)->setHotCue(a);
    (*cueAIt)->setHotCue(b);
```

Reply:
Yes, we can remove the type test.
The idea is that we can move a set hotcue to an empty slot, or swap two set hotcues.
So we need one of both cues, i.e. aborting if either of the two is missing is not correct.

Two `std::find_if()` iterations are faster than one for loop that looks for both cues?

Reply:
>  So we need only one cue with index a or b, i.e. aborting if either of the two is missing is not correct.

Ah, I see. I'll think about how to express that cleanly.

> Two std::find_if() iterations are faster than one for loop that looks for both cues?

Probably not, depends on the loop body, but its simpler IMO and probably doesn't matter here, especially if you refactor it into a private getHotcueByIndex function.